File: WoofIOS/SwiftPackages/Data/DataSource/Sources/DataSource/Dogs/DogsLocalDataSource.swift
import Combine
import DI
import DTO
import Foundation
import Utils

@SingleForProtocol
public final class DogsLocalDataSource: DogsDataSourceImplementing {
    private let scheduler: SchedulerProviding

    private static let gridLatencyMillis = 1200
    private static let profileLatencyMillis = 900

    public func getDogs(offset: Int, limit: Int) -> AnyPublisher<DogsPageDTO, DataSourceError> {
        Just(DogFixtures.page(offset: offset, limit: limit))
            .setFailureType(to: DataSourceError.self)
            .simulateNetworkLatency(millis: Self.gridLatencyMillis, scheduler: scheduler)
    }

    public func getDogProfile(dogId: Int) -> AnyPublisher<DogProfileDTO, DataSourceError> {
        Deferred { () -> AnyPublisher<DogProfileDTO, DataSourceError> in
            guard let profile = DogFixtures.profile(dogId: dogId) else {
                return Fail(error: DataSourceError.notFound(dogId: dogId)).eraseToAnyPublisher()
            }
            return Just(profile).setFailureType(to: DataSourceError.self).eraseToAnyPublisher()
        }
        .simulateNetworkLatency(millis: Self.profileLatencyMillis, scheduler: scheduler)
    }
}

private extension Publisher where Failure == DataSourceError {
    func simulateNetworkLatency(millis: Int, scheduler: SchedulerProviding) -> AnyPublisher<Output, DataSourceError> {
        delay(for: .milliseconds(millis), scheduler: scheduler.computationScheduler)
            .receive(on: scheduler.mainScheduler)
            .eraseToAnyPublisher()
    }
}

File: WoofAndroid/data/datasource/src/main/kotlin/com/perrystreet/woof/datasource/dogs/DogsLocalDataSource.kt
package com.perrystreet.woof.datasource.dogs

import com.perrystreet.woof.datasource.dogs.fixtures.DogFixtures
import com.perrystreet.woof.dto.dog.DogProfileDTO
import com.perrystreet.woof.dto.dog.DogsPageDTO
import com.perrystreet.woof.utils.scheduler.ISchedulerProvider
import io.reactivex.rxjava3.core.Single
import java.util.concurrent.TimeUnit

@org.koin.core.annotation.Single
class DogsLocalDataSource(
    private val scheduler: ISchedulerProvider,
) : IDogsDataSource {

    override fun getDogs(offset: Int, limit: Int): Single<DogsPageDTO> =
        Single.fromCallable { DogFixtures.page(offset, limit) }
            .simulateNetworkLatency(GridLatencyMillis)

    override fun getDogProfile(dogId: Long): Single<DogProfileDTO> =
        Single.fromCallable { DogFixtures.profile(dogId) }
            .simulateNetworkLatency(ProfileLatencyMillis)

    private fun <T : Any> Single<T>.simulateNetworkLatency(millis: Long): Single<T> =
        delay(millis, TimeUnit.MILLISECONDS, scheduler.computationScheduler)
            .observeOn(scheduler.mainScheduler)

    private companion object {
        const val GridLatencyMillis = 1_200L
        const val ProfileLatencyMillis = 900L
    }
}
